docs/fundamentals/code-analysis/quality-rules/ca2252.md

```diff
@@ -0,0 +1,64 @@
+---
+title: "CA2252: Opt in to Preview Features"
+description: "This analyzer scans for preview API usage and produces a build error if an assembly has not opted into preview features"
+ms.date: 09/14/2021
+ms.topic: reference
+f1_keywords:
+- CA2252
+- DetectPreviewFeaturesAnalyzer
+helpviewer_keywords:
+- DetectPreviewFeaturesAnalyzer
+- CA2252
+author: pgovind
+ms.author: pgovind
+dev_langs:
+- CSharp
+- VB
+---
+# CA2252: Opt into Preview Features before using them
+
+| | Value |
+|-|-|
+| **Rule ID** |CA2252|
+| **Category** |[Microsoft.Usage](usage-warnings.md)|
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+A client uses preview APIs/types in their assembly without explicitly opting into it either locally or at the module/assembly level.
+
+## Rule description
+
+When an API/Assembly marked with the `RequiresPreviewFeaturesAttribute` attribute is consumed, this analyzer checks if the call site has opted into preview features. A call site has opted into preview features if it is within the scope of a `RequiresPreviewFeaturesAttribute` annotation or if it is part of an assembly/module that has already opted into preview features. For ex:
+
+> ![CA2252 - Detect Preview Feature Usage](media/ca2252-basic.png)
+
+Here, `Lib` is a preview type that is constructed in the `Main` method. `Main` itself is not annotated as a preview method, so diagnostics are produced on the 2 constructors calls inside `Main`.
+
+## How to fix violations
+
+There are 2 ways to fix violations:
+
+1. Bring a call site within the scope of an annotation by marking its parent with the `RequiresPreviewFeatures` annotation. In the example above, `APreviewMethod` is annotated with the `RequiresPreviewFeatures` attribute, so the analyzer ignores preview type usage inside `APreviewMethod`. It follows that callers of `APreviewMethod` will have to perform a similar exercise.
+
+
+2. If an assembly wishes, it can also opt into preview features at an assembly or module level. This indicates to the analyzer that preview type usage in the assembly is desired and, as a consequence, no errors will produced by this analyzer. This is the preferred way to consume preview dependencies. Setting the following property in a `.csproj` file will enable preview features inside the entire assembly:
+
+```csharp
+	<PropertyGroup>
+		<EnablePreviewFeatures>true</EnablePreviewFeatures>
+	</PropertyGroup>
+```
```

Comment:
```suggestion
  ```
```

Reply:
Did you actually mean to delete the ``` here?

Reply:
Hmm, this must be a GitHub bug. The actual suggestion is to indent the ``` by two spaces.

Reply:
@gewarren Using backticks inside backticks don't always work well. Whenever I do suggestions involving backticks I use:

    ~~~suggestion
    suggestion containing ``` here
    ~~~

instead of

    ```suggestion
    suggestion containing ``` here
    ```

The following is using `~~~suggestion` instead of ``````````suggestion`

~~~suggestion
  ```
~~~

Just in case you come across this in future, it has bothered me for long before I knew about the `~~~`

Reply:
That's awesome, thanks @Youssef1313 !